### PR TITLE
[SR-510] diagnose enum raw value in closure parsing

### PR DESCRIPTION
**Pull request by dduan**
This reverts the previous fix "complain for invalid enum raw value" and fixes
SR-510 with a different approach:

~~enum raw value is parsed as a normal expression. When a closure is encountered
as the expression, a proper context will be absent. If we pass a diagnosis
from raw value to `parseExprClosure`, it will get properly issued.~~

enum raw value is parsed as a normal expression using `parseExpr()`. However,
for a closure, the parser expects a local context that doesn't exist for raw
values.

We create a temporary context to ensure the closure gets parsed as normal.
As a consequence, `parseExpr()` returns normally for closure and correct
diagnosis for raw value gets issued.

@jrose-apple

### CODE REVIEW

**Comment by jrose-apple** [comment]
Hm. First request: please split the revert out as a separate commit. :-) (Same pull request is fine, so we don't regress on master, but it is logically separate.)

Beyond that, the error message doesn't really apply here. There *is* an expression there; it's just not one that belongs there.

**Comment by dduan** [comment]
@jrose-apple done and done.

**Comment by dduan** [comment]
The error message is now "raw value for enum case must be a literal"

**Comment by jrose-apple** [comment]
This approach seems reasonable, although the hard-coding under the `case '{'` makes me a little sad.

Do you really think it's better to throw away the closure, rather than to create a dummy local context, parse it, and then complain?

**Comment by jrose-apple** [comment]
I guess we can add a very heavy assertion that *some* error was emitted if we end up in the "enum case was null" path. We can't guarantee that it happened while trying to parse the case, but if there haven't been any *other* errors, we won't let this one slip by.

**Comment by dduan** [comment]
>  the hard-coding under the case '{' makes me a little sad.

I don't feel good about that neither, it's a hack at best.

> Do you really think it's better to throw away the closure, rather than to create a dummy local context, parse it, and then complain?

We want to complain about the raw value being a closure, not what's wrong in the closure itself, no?

> I guess we can add a very coarse assertion that some error was emitted if we end up in the "enum case was null" path.

That's basically what's in master now. For `case = {}`, we get both "expect expression after =" and "non-literal" diagnosis (not these exact phrases, obviously).

> …but if there haven't been any other errors…

I couldn't figure out how to do this. Tips?

A couple of scenarios with this change applied:

```
$ cat /tmp/t.swift
enum X: String {
    case A
    case B = {
$ ./swift /tmp/t.swift
/tmp/t.swift:3:14: error: raw value for enum case must be a literal
    case B = {
             ^
/tmp/t.swift:4:1: error: expected declaration
```

```
$ cat /tmp/t.swift
enum X: String {
    case A
    case B =
$ ./swift /tmp/t.swift
/tmp/t.swift:3:13: error: expected expression after '=' in 'case'
    case B =
            ^
/tmp/t.swift:4:1: error: expected declaration

^
```

**Comment by dduan** [comment]
@jrose-apple the cosmetic changes has been made.

**Comment by dduan** [comment]
@jrose-apple `is` only exists for `swift::Diagnostic`, so I can't use it on `swift::Diag` 🙁

**Comment by dduan** [comment]
@jrose-apple running test on coarse assertion for null raw value now.

**Comment by dduan** [comment]
I tried adding an assertion to raw value's null branch. It caused a few assertion failures in some tests regarding name binding.

**Comment by dduan** [comment]
Ok, assertion added.

**Comment by dduan** [comment]
Update for @jrose-apple:

I implemented your suggestion. Adding a local context made the patch much simpler. The closure gets parsed properly and errors related to both the closure itself and raw value gets issued properly.

An assertion for the "enum case was null" path is also added.

All tests are passing.

File: lib/Parse/ParseDecl.cpp [hunk 1]
```diff
@@ -4413,7 +4412,7 @@ ParserStatus Parser::parseDeclEnumCase(ParseDeclOptions Flags,
         return Status;
       }
       if (RawValueExpr.isNull()) {
-        diagnose(NextLoc, diag::nonliteral_enum_case_raw_value);
+        assert("No error emmited for invalid enum raw value expression");
```
Comment (nuclearace) [inline lib/Parse/ParseDecl.cpp]: ~~emmited~~ emitted?

Comment (dduan) [inline lib/Parse/ParseDecl.cpp]: Thank Erik. Fixed.

**Comment by jrose-apple** [comment]
Um. We can work out the assertion failures separately (it still seems valuable). I actually meant creating a local context ahead of time, so that we wouldn't have to hardcode a particular ID in the main expr-parsing loop. I guess this is cheaper, though, since it only comes up on failure. Would it at least make sense to move it into the closure-parsing logic, though?

Also, how does that local context ever get cleaned up? Seems like it's just new'ed into space.

**Comment by dduan** [comment]
@jrose-apple Ok, I removed the hard-coded diagnosis from `parseExpr`, tests are parsing.

**Comment by jrose-apple** [comment]
That still leaks the local context, though, doesn't it?

**Comment by dduan** [comment]
@jrose-apple Oops, fixed.

**Comment by dduan** [comment]
📣

**Comment by dduan** [comment]
@jrose-apple anything else I can do to make this one better?

**Comment by jrose-apple** [comment]
Sorry I hadn't gotten back to this! I think this is pretty good. It's not a completely beautiful solution but it does mean we're parsing the same way as everything else. I have nitpicks on your comments but I'll merge it after that.

**Comment by dduan** [comment]
Done. Thank you for all the help 😬.

**Pull request merged**